File: src/main/java/ws/gross/gradle/utils/GradleUtils.java
/*
 * Copyright 2023 Konstantin Gribov
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ws.gross.gradle.utils;

import java.io.IOException;
import java.io.Reader;
import java.io.Serializable;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import javax.annotation.Nullable;

import org.gradle.api.Action;
import org.gradle.api.GradleException;
import org.gradle.api.Project;
import org.gradle.api.artifacts.dsl.RepositoryHandler;
import org.gradle.api.artifacts.repositories.ArtifactRepository;
import org.gradle.api.artifacts.repositories.MavenArtifactRepository;
import org.gradle.api.credentials.Credentials;
import org.gradle.api.provider.Provider;
import org.gradle.authentication.http.BasicAuthentication;
import org.gradle.internal.artifacts.repositories.AuthenticationSupportedInternal;

public class GradleUtils {
  private static final Action<?> DO_NOTHING = new NoopAction<>();
  private static final Pattern SPLIT_COMMA_PATTERN = Pattern.compile(",");
  private static final Pattern SPLIT_EQUALS_PATTERN = Pattern.compile("=");

  private GradleUtils() {
  }

  @SuppressWarnings("unchecked")
  public static <T> Action<T> doNothing() {
    return (Action<T>) DO_NOTHING;
  }

  public static MavenArtifactRepository maven(
      RepositoryHandler repositories,
      String repoName,
      Provider<String> repoUrl,
      Provider<? extends Credentials> credentials
  ) {
    return maven(repositories, repoName, repoUrl, credentials, doNothing());
  }

  public static MavenArtifactRepository maven(
      RepositoryHandler repositories,
      String repoName,
      Provider<String> repoUrl,
      Provider<? extends Credentials> credentials,
      Action<? super MavenArtifactRepository> configureAction
  ) {
    ArtifactRepository repo = repositories.findByName(repoName);
    if (repo instanceof MavenArtifactRepository) {
      return (MavenArtifactRepository) repo;
    } else if (repo != null) {
      throw new GradleException("Repository with name " + repoName + " is not a MavenArtifactRepository");
    } else {
      return repositories.maven(r -> {
        r.setName(repoName);
        r.setUrl(repoUrl);
        if (credentials.isPresent()) {
          r.authentication(a -> a.create("basic", BasicAuthentication.class));
          ((AuthenticationSupportedInternal) r).getConfiguredCredentials().set(credentials);
        }
        configureAction.execute(r);
      });
    }
  }

  public static Properties readProperties(Path path) {
    try (Reader reader = Files.newBufferedReader(path, StandardCharsets.UTF_8)) {
      Properties result = new Properties();
      result.load(reader);
      return result;
    } catch (IOException e) {
      throw new GradleException("Failed to read " + path, e);
    }
  }

  public static List<String> parseList(@Nullable String value) {
    return SPLIT_COMMA_PATTERN.splitAsStream(value == null ? "" : value)
        .map(String::trim)
        .filter(s -> !s.isEmpty())
        .collect(Collectors.toList());
  }

  public static Map<String, String> parseMap(@Nullable String value) {
    return parseList(value)
        .stream()
        .map(p -> {
          String[] parts = SPLIT_EQUALS_PATTERN.split(p);
          if (parts.length != 2) {
            throw new IllegalArgumentException("Unexpected pair " + p + " while parsion map");
          }
          return parts;
        })
        .collect(Collectors.toMap(p -> p[0], p -> p[1]));
  }

  public static boolean isDslAccessorsGeneration(Project project) {
    return project.getName().equals("gradle-kotlin-dsl-accessors");
  }

  public static class NoopAction<T> implements Action<T>, Serializable {
    @Override
    public void execute(T _ignored) {
    }
  }
}

File: src/main/kotlin/ws/gross/gradle/gradle-utils.kt
/*
 * Copyright 2022 Konstantin Gribov
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ws.gross.gradle

import org.gradle.api.Action
import org.gradle.api.GradleException
import org.gradle.api.Project
import org.gradle.api.artifacts.MutableVersionConstraint
import org.gradle.api.artifacts.VersionConstraint
import org.gradle.api.artifacts.dsl.DependencyHandler
import org.gradle.api.initialization.Settings
import org.gradle.api.initialization.dsl.VersionCatalogBuilder
import org.gradle.api.internal.FeaturePreviews
import org.gradle.api.provider.Provider
import org.gradle.kotlin.dsl.support.serviceOf
import org.gradle.plugin.use.PluginDependency
import org.gradle.util.GradleVersion
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType

inline fun Project.ifNotDslAccessors(block: () -> Unit) {
  if (project.name != "gradle-kotlin-dsl-accessors") block.invoke()
}

// See https://github.com/gradle/gradle/issues/18620
fun DependencyHandler.plugin(pluginId: String, version: String) =
  create("$pluginId:$pluginId.gradle.plugin:$version")
@Suppress("UnstableApiUsage")
fun DependencyHandler.plugin(plugin: Provider<PluginDependency>) =
  plugin.get().run { plugin(pluginId, version.displayName) }

fun String.parsePublishTaskInfo(): PublishTaskInfo? =
  publishTaskNameRegex.matchEntire(this)?.destructured?.let { (p, r) ->
    PublishTaskInfo(publication = p.decapitalize(), repository = r.decapitalize())
  }

data class PublishTaskInfo(val publication: String, val repository: String)

fun isVersionCatalogsSupported() = GRADLE_7_0_PLUS
fun isVersionCatalogsExperimental(): Boolean {
  if (!isVersionCatalogsSupported()) {
    throw GradleException("version catalogs not supported")
  }
  return VersionCatalogsUtil.featureActive
}

fun Settings.isVersionCatalogsEnabled() = GRADLE_7_4_PLUS
    || (isVersionCatalogsSupported() && VersionCatalogsUtil.isFeatureEnabled(this))

fun MutableVersionConstraint.fromConstraint(v: VersionConstraint) {
  if (v.preferredVersion.isNotEmpty()) prefer(v.preferredVersion)
  if (v.requiredVersion.isNotEmpty()) require(v.requiredVersion)
  if (v.strictVersion.isNotEmpty()) strictly(v.strictVersion)
  if (v.rejectedVersions.isNotEmpty()) reject(*v.rejectedVersions.toTypedArray())
}

internal object ProviderUtil {
  @JvmStatic
  private val FOR_USE_AT_CONFIGURATION_TIME = try {
    MethodHandles.publicLookup().findVirtual(
      Provider::class.java,
      "forUseAtConfigurationTime",
      MethodType.methodType(Provider::class.java)
    )
  } catch (e: NoSuchMethodException) {
    null
  }

  @Suppress("UNCHECKED_CAST")
  fun <T> Provider<T>.wrappedForUseAtConfigurationTime(): Provider<T> = if (GRADLE_7_4_PLUS) {
    this
  } else {
    FOR_USE_AT_CONFIGURATION_TIME!!.invokeExact(this) as Provider<T>
  }
}

internal object VersionCatalogsUtil {
  val featureActive = FeaturePreviews.Feature.VERSION_CATALOGS.isActive

  fun isFeatureEnabled(settings: Settings) =
    settings.serviceOf<FeaturePreviews>().isFeatureEnabled(FeaturePreviews.Feature.VERSION_CATALOGS)
}

private val publishTaskNameRegex = """publish([A-Z][\w.]*)PublicationTo([A-Z]\w*)Repository""".toRegex()

private val GRADLE_7_4_PLUS = GradleVersion.current() >= GradleVersion.version("7.4")
private val GRADLE_7_0_PLUS = GradleVersion.current() >= GradleVersion.version("7.0")
